src/main/java/edu/byu/cs/autograder/git/specs.md

```diff
@@ -0,0 +1,37 @@
+
+## Local Clocks
+This also opens the door to issues relating to the proper setting of a student’s _local clock_ since the timestamps of commits are based on the local time of the computer; this means a student could cheese the system by changing the local time on their computer to a different day before committing to still meet the requirements. One way to verify this is to verify that all commits are authored in a strictly chronological order, _and that the previous head hash is a direct ancestor of the current submission_.
+
+Besides setting their clocks backwards, a student could also set their clock forward to gain extra days. If a student’s clock is set in the future, then they could easily fast forward an extra day before authoring their next commit, and then fast-forward the clock and commit again. This would count as two commits on two days and would get them their credit.
+
+## Advanced `git` commands
+However, I just realized that strictly tracking the commit hashes would also not be foolproof since a student could easily rebase their code and force-push their branch. These commits would all be “new” and count towards their history even though they really are not. Likely, no student in this class is even using multiple branches or even know how to `git rebase` as these are certainly advanced commands. Additionally, even if a student knew how to do the commands, they likely won’t realize that this is a vulnerability and a way to cheese the system.
+
+## Recommendation
+Based on all the information above, I propose the following:
+1. We count only commits towards the commit limit if they were authored after **the most recent timestamp of the commit of any commit that resulted in an accepted submission from any phase**.
+2. We throw an error and refuse to grade if we detect any commits that are authored before their parent commit (indicates intentionally messing with the local clock, or cherry-picking commits).
+3. We withhold the score as “not qualified” if any of the commits are authored in the future of the server timestamp (indicates messing with the local clock, or it could be a simple mistake).
+4. We explicitly preserve a student’s commit verification result for any subsequent submissions on the same phase. I.E. If they pass on their first valid submission, then we’ll continue passing them even if they are resubmitting to fix quality errors etc… And, if they fail on their first submissions, they’ll be able to continue resubmitting, but none of those grades will be submitted to Canvas even if they are resolving quality errors.
```

Comment:
I agree that this is the only way to fight against fake commits and should be implemented, but this could cause a very confusing experience for users. There will need to be very clear instruction both in class, on GitHub, and on the autograder explaining this rule

Reply:
I've done my best to provide helpful error messages. For example, the autograder will report the following in the notes which go to Canvas:

> Commit history approved with a penalty of 10%